apps/alb_monitor/lib/alb_monitor/monitor.ex

```diff
@@ -0,0 +1,94 @@
+defmodule ALBMonitor.Monitor do
+  @moduledoc "See `ALBMonitor`."
+
+  use GenServer
+
+  alias ExAws.ElasticLoadBalancingV2, as: ALB
+
+  require Logger
+
+  @ex_aws Application.compile_env!(:alb_monitor, :ex_aws)
+  @http Application.compile_env!(:alb_monitor, :http)
+
+  defmodule State do
+    @moduledoc "State of the GenServer."
+    defstruct check_interval: 5_000,
+              fetch_env_fn: &System.fetch_env/1,
```

Comment:
I think I see what you mean, but how would tests fake and manipulate the environment with this approach?

Reply:
You could either update the environment, or have it overridable:

```elixir
defmodule State
  defstruct arn: Application.get_env!(:alb_monitor, :arn)
end

def start_link(overrides // %{}) do
  state = struct!(State, overrides)
  GenServer.start_link(__MODULE__, state)
end

# in test
start_link(arn: "test_arn")
```

Reply:
```elixir
defstruct arn: Application.fetch_env!(:alb_monitor, :arn)
```

I'm not sure this would work, since it would be getting the value at compile time, but the value wouldn't be present until runtime. It would have to be like:

```elixir
defstruct arn_fn: fn -> Application.fetch_env!(:alb_monitor, :arn) end
```

And then you're pretty much back to the `fetch_env_fn` approach, but with more `fn`s. Though there might be value in having all the runtime environment variables nicely listed in `releases.exs`; in that case it could work like this?

```elixir
defstruct env_fn: &Application.fetch_env!(:alb_monitor, &1)

# in releases.exs
config :alb_monitor, target_group_arn: System.get_env("ALB_TARGET_GROUP_ARN"), ...
```

Reply:
I thought they happened at runtime, but if not, you can have it as a default in `start_link`:
```elixir
def start_link(overrides // %{}) do
  overrides = Map.put_new(overrides, :arn, Application.fetch_env!(:alb_monitor, :arn))
  ...
end

Reply:
Tried doing it this way (more or less) in my latest commit. I'm in favor of putting the env vars in `releases.exs`, and it did allow extraction of the `no_ecs_metadata_uri` case to a separate function clause, which is nice.